corona-warn-app/cwa-app-android
# Perform data reset for users already affected by the encryption issues (EXPOSUREAPP-1851/EXPOSUREAPP-2958)

~~**DO NOT MERGE** but review :wink:~~
~~This still needs the final translations for the info dialog after a reset.~~

## Description
The mitigation from #1235 only helps users not already affected by the issue (see #642).

Existing users will still keep crashing. There is graceful recovery because the encryption key is gone.
They have to reset their app data. We can do this for them. If the user has automatic updates enabled in Google Play, there may be users who have already been affected, don't know it yet, and where it will be fixed automatically too. They will only see a dialog when they open their app the next time, informing them about what happened (see mock screenshot).

The result should be that the app automatically starts working again if an affected user updates it. We will loose the encrypted data, but the automatic tasks should be able to work again and thus the app  perform it's duty.

**When do we do it?**
* It is a `GeneralSecurityException("decryption failed")`
* The encrypted shared preferences actually exist (otherwise it's a different error cause)
* This is the first error we encountered when creating the encrypted shared preferences, which means that there was no other error before we encountered this one. This allows us to only perform the reset for users upgrading from <1.4.0 to 1.4.0+, as we don't want to reset for any other case potentially hiding other issues.

**What do we do?**
* If the specific error case is matched
* Deleted the encrypted shared preferences
* Delete the encrypted database
* Retry creation of the encrypted shared preferences

## How to test
* Open the app
* Review the code
* Review the tests, are all edge cases covered?
* Install the app, complete onboarding, insert a `throw GeneralSecurityException("decryption failed")` ~~into the `EncryptedPreferencesFactory`~~ actually, insert it into the [SecurityHelper](https://github.com/corona-warn-app/cwa-app-android/pull/1256/files#diff-d05886a3ee6f8f08529a36f8e9fb936bR49), otherwise we keep throwing. Install the new version without data reset. You should see the app reset and then display the dialog, once.

<img src="https://user-images.githubusercontent.com/1439229/94567688-06e00400-026c-11eb-83c6-3f303b857766.png" width="400">
